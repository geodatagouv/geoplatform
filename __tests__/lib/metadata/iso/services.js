const {getWFSServiceLocation, getCoupledResources} = require('../../../../lib/metadata/iso/services')

describe('lib.metadata.iso.services', () => {
  describe('getWFSServiceLocation()', () => {
    it('should return the service URL', () => {
      const metadata = {
        distributionInfo: {
          transferOptions: [
            {
              onLine: [
                {
                  linkage: 'https://example.com/wfs?SERVICE=WFS&REQUEST=GetCapabilities&',
                  protocol: 'OGC:WFS-2.0.0-http-get-capabilities'
                }
              ]
            }
          ]
        }
      }

      expect(getWFSServiceLocation(metadata)).toBe('https://example.com/wfs')
    })

    it('should return null if there are no WFS service detected', () => {
      const metadata = {
        distributionInfo: {
          transferOptions: [
            {
              onLine: [
                {
                  linkage: 'https://example.com/random'
                }
              ]
            }
          ]
        }
      }

      expect(getWFSServiceLocation(metadata)).toBeNull()
    })

    it('should return null if the WFS service is not a valid url', () => {
      const testCases = [
        'https://example.unknowntld/wfs?SERVICE=WFS&REQUEST=GetCapabilities',
        'https://127.0.0.1/wfs?SERVICE=WFS&REQUEST=GetCapabilities',
        '/wfs?SERVICE=WFS&REQUEST=GetCapabilities',
        'ftp://example.com/wfs?SERVICE=WFS&REQUEST=GetCapabilities',
        'foobar'
      ]

      for (const linkage of testCases) {
        expect(getWFSServiceLocation({
          distributionInfo: {
            transferOptions: [
              {
                onLine: [
                  {
                    linkage,
                    protocol: 'OGC:WFS-2.0.0-http-get-capabilities'
                  }
                ]
              }
            ]
          }
        })).toBeNull()
      }
    })

    it('should return null if there are multiple WFS services listed', () => {
      const metadata = {
        distributionInfo: {
          transferOptions: [
            {
              onLine: [
                {
                  linkage: 'https://example.com/wfs-1?SERVICE=WFS&REQUEST=GetCapabilities&',
                  protocol: 'OGC:WFS-2.0.0-http-get-capabilities'
                },
                {
                  linkage: 'https://example.com/wfs-2?SERVICE=WFS&REQUEST=GetCapabilities&',
                  protocol: 'OGC:WFS-2.0.0-http-get-capabilities'
                }
              ]
            }
          ]
        }
      }

      expect(getWFSServiceLocation(metadata)).toBeNull()
    })
  })

  describe('getCoupledResources()', () => {
    it('should return the coupled resources', () => {
      const metadata = {
        identificationInfo: {
          coupledResource: [
            {
              operationName: 'GetFeature',
              identifier: 'foo1',
              scopedName: 'feature_type-1'
            },
            {
              operationName: 'GetFeature',
              identifier: 'foo2',
              scopedName: 'feature_type-2'
            }
          ]
        }
      }

      expect(getCoupledResources(metadata)).toMatchSnapshot()
    })

    it('should dedup duplicated coupled resources', () => {
      const metadata = {
        identificationInfo: {
          coupledResource: [
            {
              operationName: 'GetCapabilities',
              identifier: '9fd801e3-976a-49b6-b84a-0bd448fec8d5',
              scopedName: 'bati_avap_paimpol'
            },
            {
              operationName: 'DescribeFeatureType',
              identifier: '9fd801e3-976a-49b6-b84a-0bd448fec8d5',
              scopedName: 'bati_avap_paimpol'
            },
            {
              operationName: 'GetFeature',
              identifier: '9fd801e3-976a-49b6-b84a-0bd448fec8d5',
              scopedName: 'bati_avap_paimpol'
            },
            {
              operationName: 'GetCapabilities',
              identifier: 'e3b34f32-8081-4451-83e0-4d5fcfd36512',
              scopedName: 'elr_avap_paimpol'
            },
            {
              operationName: 'DescribeFeatureType',
              identifier: 'e3b34f32-8081-4451-83e0-4d5fcfd36512',
              scopedName: 'elr_avap_paimpol'
            }
          ]
        }
      }

      expect(getCoupledResources(metadata)).toMatchSnapshot()
    })

    it('should ignore incomplete resources', () => {
      const metadata = {
        identificationInfo: {
          coupledResource: [
            {
              operationName: 'GetFeature',
              identifier: 'foo1'
            },
            {
              identifier: 'foo2',
              scopedName: 'feature_type-2'
            },
            {
              scopedName: 'feature_type-3'
            }
          ]
        }
      }

      expect(getCoupledResources(metadata)).toMatchSnapshot()
    })

    it('should return an empty array if there are no coupled resources', () => {
      const testCases = [
        [],
        null,
        undefined
      ]

      for (const coupledResource of testCases) {
        expect(getCoupledResources({
          identificationInfo: {
            coupledResource
          }
        })).toEqual([])
      }
    })
  })
})
