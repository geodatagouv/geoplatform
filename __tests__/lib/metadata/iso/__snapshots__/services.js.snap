// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib.metadata.iso.services getCoupledResources() should dedup duplicated coupled resources 1`] = `
Array [
  Object {
    "identifier": "9fd801e3-976a-49b6-b84a-0bd448fec8d5",
    "scopedName": "bati_avap_paimpol",
  },
  Object {
    "identifier": "e3b34f32-8081-4451-83e0-4d5fcfd36512",
    "scopedName": "elr_avap_paimpol",
  },
]
`;

exports[`lib.metadata.iso.services getCoupledResources() should ignore incomplete resources 1`] = `
Array [
  Object {
    "identifier": "foo2",
    "scopedName": "feature_type-2",
  },
]
`;

exports[`lib.metadata.iso.services getCoupledResources() should return the coupled resources 1`] = `
Array [
  Object {
    "identifier": "foo1",
    "scopedName": "feature_type-1",
  },
  Object {
    "identifier": "foo2",
    "scopedName": "feature_type-2",
  },
]
`;
